packages/language-service/lib/service-plugin.js

```diff
@@ -0,0 +1,67 @@
+/**
+ * @typedef {import('@volar/language-service').ServicePlugin} ServicePlugin
+ */
+
+import {fromPlace} from 'unist-util-lsp'
+import {VirtualMdxFile} from './virtual-file.js'
+
+/**
+ * Create an Volar service plugin for MDX files.
+ *
+ * The service supports reporting diagnostics for parsing errors.
+ *
+ * @returns {ServicePlugin}
+ *   The Volar service plugin for MDX files.
+ */
+export function createMdxServicePlugin() {
+  return {
+    create(context) {
+      return {
+        provideSemanticDiagnostics(document) {
+          const file = getVirtualMdxFile(document.uri)
+
+          const error = file?.error
+
+          if (error) {
+            return [
+              {
+                message: error.message,
+                code: error.ruleId,
+                codeDescription: {
+                  href:
+                    error.url || 'https://mdxjs.com/docs/troubleshooting-mdx/'
```

Comment:
Some vfile messages contain a URL. These can link to for example https://github.com/micromark/micromark-extension-mdxjs-esm#could-not-parse-importexports-with-acorn. The MDX website also has a nice troubleshooting guide at https://mdxjs.com/docs/troubleshooting-mdx/.

Both are nice, but I have mixed feelings about sending the user to different pages. Perhaps it would be nicer for MDX related parsing errors to always link to a page in the MDX troubleshooting docs.

This isn’t blocking. The URL can be changed later.

@wooorm @ChristianMurphy WDYT?

Reply:
I'm fine with there being a fallback.
PRs giving more specific guides to MDX errors are welcome

Reply:
sorry looks like my comment didn’t go through, it was pending, just submitted

Reply:
I think the fallback is really useful. It contains a very well-written list of potential errors. I just think it’s a bit weird to link to a micromark plugin. I understand that errors come from these plugins, but as an end user this feels like an implementation detail. Nevertheless, it is great to point directly to a section that explains the exact error. It might be nice for micromark extension errors to point to the corresponding error on https://mdxjs.com/docs/troubleshooting-mdx/ instead.

Re `error.source`: Good idea, I added support for it in 77c76511b0f64afc2a9525aa031f4c66fded0b01
Re severity: The severity is 1 / Error, which is the most severe.

Reply:
There are other potential errors, many things could fail, folks will be sent to the wrong place.

The trouble shooting guide works well with Google. If you search for an error message Google will show the snippet already. Probably other search engines too.

I don’t think micromark extension should point to MDX; it might be combined with other things. Say, Astro. To create a custom format unrelated to MDX.
A better solution, I believe, is that `ruleId`s can be used by higher level tools to map to different URLs. I’m open to doing that in `@mdx-js/mdx` as well.